openmc/summary.py

```diff
@@ -135,365 +117,245 @@ def _read_materials(self):
+                cell.temperature = group['temperature'][...]
 
             # Store Cell fill information for after Universe/Lattice creation
-            self._cell_fills[index] = (fill_type, fill)
+            cell_fills[cell.id] = (fill_type, fill)
 
             # Generate Region object given infix expression
             if region:
-                cell.region = Region.from_expression(
-                    region, {s.id: s for s in self.surfaces.values()})
+                cell.region = Region.from_expression(region, surfaces)
 
             # Get the distribcell data
-            if 'distribcell_index' in self._f['geometry/cells'][key]:
-                ind = self._f['geometry/cells'][key]['distribcell_index'].value
+            if 'distribcell_index' in group:
+                ind = group['distribcell_index'].value
                 cell.distribcell_index = ind
-                paths = self._f['geometry/cells'][key]['paths'][...]
+                paths = group['paths'][...]
                 paths = [str(path.decode()) for path in paths]
                 cell.distribcell_paths = paths
 
             # Add the Cell to the global dictionary of all Cells
-            self.cells[index] = cell
+            cells[cell.id] = cell
 
-    def _read_universes(self):
-        self.n_universes = self._f['geometry/n_universes'].value
+        return cells, cell_fills
 
-        # Initialize dictionary for each Universe
-        # Keys     - Universe keys
-        # Values   - Universe objects
-        self.universes = {}
+    def _read_universes(self, cells):
+        universes = {}
 
         for key in self._f['geometry/universes'].keys():
-            if key == 'n_universes':
-                continue
-
             universe_id = int(key.lstrip('universe '))
-            index = self._f['geometry/universes'][key]['index'].value
-            cells = self._f['geometry/universes'][key]['cells'][...]
+            cell_ids = self._f['geometry/universes'][key]['cells'][...]
 
             # Create this Universe
             universe = openmc.Universe(universe_id=universe_id)
 
             # Add each Cell to the Universe
-            for cell_id in cells:
-                cell = self.cells[cell_id]
-                universe.add_cell(cell)
+            for cell_id in cell_ids:
+                universe.add_cell(cells[cell_id])
 
             # Add the Universe to the global list of Universes
-            self.universes[index] = universe
-
-    def _read_lattices(self):
-        self.n_lattices = self._f['geometry/n_lattices'].value
+            universes[universe.id] = universe
 
-        # Initialize lattices for each Lattice
-        # Keys     - Lattice keys
-        # Values   - Lattice objects
-        self.lattices = {}
+        return universes
 
-        for key in self._f['geometry/lattices'].keys():
-            if key == 'n_lattices':
-                continue
+    def _read_lattices(self, universes):
```

Comment:
I did it for the cases where it's obvious that there's no extra information needed. For example, to have a `Cell.from_hdf5` method, you also potentially need material and universe information (to assign the fill). However, at the time we're reading cells, we don't have universe information yet. All that information gets combined in the `_finalize_geometry` method. I can revisit this though and see if it might be feasible by changing the order of things.